src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,6 +177,21 @@ private bool CompleteCore()
             {
                 if (existingWaiters != null)
                 {
+                    // we have waiters and we have data.
+                    // allow some extra time to drain the queue before cancelling
+                    // but do not wait for too long.
+                    SpinWait.SpinUntil(() =>
+                    {
+                        lock (SyncObject)
+                        {
+                            return _data.Count == 0;
+                        }
+                    },
+                    10000);
```

Comment:
Yes. Generally we would not need to wait, so the number is not hugely important.
I was suggested to use 30sec. here, but decided that it is too long. Typical compile of a project the size of C# compiler is 3-6 sec. Depending on machine. So I think 10sec is enough for now.

We should try find a better solution though, if possible. That is tracked by https://github.com/dotnet/roslyn/issues/11470